File: karaoke/services/queue_service.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os

from tortoise.exceptions import DoesNotExist

from karaoke.domain.playback import persist_playback_mode, resolve
from karaoke.domain.prepare_policy import profile_needs_prepare
from karaoke.domain.queue_policy import QueueState
from karaoke.dto.mappers import history_item
from karaoke.errors import fail_result
from karaoke.events.bus import event_bus
from karaoke.infra.repositories.history_repo import HistoryRepository
from karaoke.infra.repositories.song_repo import SongRepository
from karaoke.dto.api_result import ApiResult
from karaoke.services.base import apply_pagination, run_guarded
from karaoke.services.prepare_service import PrepareService
from settings import PAGE_SIZE


class QueueService:
    def __init__(
        self,
        songs: SongRepository = None,
        histories: HistoryRepository = None,
        prepare: PrepareService = None,
    ) -> None:
        self._songs = songs or SongRepository()
        self._histories = histories or HistoryRepository()
        self._prepare = prepare or PrepareService()

    async def init_on_startup(self) -> None:
        await self._histories.reset_stale_singing()

    async def _build_list(self, histories) -> list:
        if not histories:
            return []
        song_map = await self._songs.map_by_ids([h.id for h in histories])
        return [history_item(h, song_map.get(h.id)) for h in histories]

    async def enqueue(self, song_id: int) -> ApiResult:
        result = ApiResult()
        try:
            song = await self._songs.get(song_id)
            profile = resolve(song)
            if not profile.can_queue:
                result.code = 1
                if not song.is_playable or not os.path.isfile(song.source_path):
                    result.msg = "源视频不可播放或不存在"
                else:
                    result.msg = "增强资源不完整且源视频不可用"
                return result

            if profile_needs_prepare(song, profile):
                prep_status = await self._prepare.status(song.id)
                if not prep_status.get('ready'):
                    prep = await self._prepare.schedule(song.id)
                    result.code = 1
                    result.msg = "播放资源正在准备中，请耐心等待"
                    result.data = {'prepare': prep}
                    return result

            history = await self._histories.get_optional(song.id)
            if history:
                if history.is_sing == QueueState.SUNG:
                    history.is_sing = QueueState.PENDING
                    history.is_top = 0
                    await self._histories.save(history, ['is_sing', 'is_top', 'update_time'])
            else:
                await self._histories.create(
                    id=song.id, name=song.display_name,
                    is_sing=QueueState.PENDING, is_top=0,
                )

            await persist_playback_mode(song, profile)

            await event_bus.publish_queue_changed()
            result.msg = f"{song.display_name} 点歌成功"
            result.data = {'playback_mode': profile.mode}
        except DoesNotExist:
            result.code = 1
            result.msg = "歌曲不存在"
        except Exception as exc:
            fail_result(result, exc, "点歌失败")
        return result

    async def list_pending(self) -> ApiResult:
        return await self._list_by_type('pendingAll')

    async def list_history(self, page: int = 1) -> ApiResult:
        return await self._list_by_type('history', page)

    async def list_usually(self, page: int = 1) -> ApiResult:
        return await self._list_by_type('usually', page)

    async def _list_by_type(self, query_type: str, page: int = 1) -> ApiResult:
        async def load():
            total_num = 0
            if query_type == 'history':
                histories, total_num = await self._histories.list_history_page(page)
            elif query_type == 'usually':
                histories, total_num = await self._histories.list_usually_page(page)
            elif query_type == 'pendingAll':
                histories = await self._histories.list_pending()
                total_num = len(histories)
            else:
                return ApiResult.fail(f"未知查询类型: {query_type}")
            data = await self._build_list(histories)
            result = ApiResult(data=data)
            apply_pagination(result, total_num, page, PAGE_SIZE)
            if query_type == 'pendingAll':
                result.totalPage = 1 if total_num else 0
            return result

        return await run_guarded('获取队列失败', load)

    async def set_top(self, song_id: int) -> ApiResult:
        result = ApiResult()
        try:
            history = await self._histories.get(song_id)
            history.is_top = 1
            await self._histories.save(history, ['is_top', 'update_time'])
            result.msg = f"{history.name} 置顶成功"
            await event_bus.publish_queue_changed()
        except DoesNotExist:
            result.code = 1
            result.msg = "歌曲不在队列中"
        except Exception as exc:
            fail_result(result, exc, "置顶失败")
        return result

    async def remove_if_exists(self, song_id: int) -> None:
        history = await self._histories.get_optional(song_id)
        if history:
            await self._histories.delete(history)

    async def remove(self, song_id: int) -> ApiResult:
        result = ApiResult()
        try:
            history = await self._histories.get(song_id)
            await self._histories.delete(history)
            result.msg = f"{history.name} 播放记录删除成功"
            await event_bus.publish_queue_changed()
        except DoesNotExist:
            result.code = 1
            result.msg = "歌曲不在队列中"
        except Exception as exc:
            fail_result(result, exc, "移除队列项失败")
        return result

File: crates/services/src/queue_service.rs
//! 点歌队列服务。对应 Python `karaoke/services/queue_service.py`。

use crate::dto::{db_error_message, ApiResult};
use crate::mappers::history_item;
use karaoke_domain::playback::effective_mode;
use karaoke_domain::queue_policy::QueueState;
use karaoke_events::EventBus;
use karaoke_infra::repositories::history_repo::PAGE_SIZE;
use karaoke_infra::repositories::{HistoryRepository, SongRepository};
use karaoke_infra::PlaybackResolver;
use karaoke_jobs::PrepareTaskManager;
use std::sync::Arc;

#[derive(Clone)]
pub struct QueueService {
    pub songs: SongRepository,
    pub histories: HistoryRepository,
    pub resolver: PlaybackResolver,
    pub prepare: Arc<PrepareTaskManager>,
    pub events: EventBus,
}

impl QueueService {
    pub async fn init_on_startup(&self) {
        match self.histories.reset_stale_singing().await {
            Ok(n) if n > 0 => tracing::info!("reset {n} stale singing rows on startup"),
            Ok(_) => {}
            Err(e) => tracing::warn!("reset_stale_singing failed: {e}"),
        }
    }

    async fn build_list(
        &self,
        histories: &[karaoke_infra::models::HistoryRow],
    ) -> Vec<crate::mappers::HistoryItem> {
        if histories.is_empty() {
            return vec![];
        }
        let ids: Vec<i64> = histories.iter().map(|h| h.id).collect();
        let song_map = self.songs.map_by_ids(&ids).await.unwrap_or_default();
        histories
            .iter()
            .map(|h| history_item(h, song_map.get(&h.id)))
            .collect()
    }

    pub async fn enqueue(&self, song_id: i64) -> ApiResult {
        let song = match self.songs.get_optional(song_id).await {
            Ok(Some(s)) => s,
            Ok(None) => return ApiResult::not_found("歌曲"),
            Err(e) => return ApiResult::fail(db_error_message(&e, "点歌失败")),
        };

        let profile = self.resolver.resolve(&song).await;
        if !profile.can_queue {
            let msg = if !song.is_playable || !std::path::Path::new(&song.source_path).is_file() {
                "源视频不可播放或不存在"
            } else {
                "增强资源不完整且源视频不可用"
            };
            return ApiResult::fail(msg);
        }

        if self.resolver.needs_prepare(&song, &profile).await {
            let prep_status = self.prepare.status(song_id).await;
            if !prep_status.ready {
                let prep = self.prepare.schedule(song_id).await;
                return ApiResult::fail_with_data(
                    "播放资源正在准备中，请耐心等待",
                    serde_json::json!({ "prepare": prep }),
                );
            }
        }

        match self.histories.get_optional(song_id).await {
            Ok(Some(history)) => {
                if history.is_sing == QueueState::Sung.to_db() {
                    if let Err(e) = self.histories.set_pending(song_id).await {
                        return ApiResult::fail(db_error_message(&e, "点歌失败"));
                    }
                }
            }
            Ok(None) => {
                if let Err(e) = self
                    .histories
                    .create(song_id, &song.display_name, QueueState::Pending.to_db(), 0)
                    .await
                {
                    return ApiResult::fail(db_error_message(&e, "点歌失败"));
                }
            }
            Err(e) => return ApiResult::fail(db_error_message(&e, "点歌失败")),
        }

        if let Err(e) = self
            .songs
            .update_playback_meta(
                song_id,
                effective_mode(&profile, &song.playback_mode).as_str(),
                Some(profile.playback_source.as_str()),
                profile.can_queue,
            )
            .await
        {
            tracing::warn!("persist playback meta failed for song {song_id}: {e}");
        }

        self.events.publish_queue_changed();
        let display = if song.display_name.trim().is_empty() {
            format!("歌曲 #{song_id}")
        } else {
            song.display_name.clone()
        };
        ApiResult::ok_msg_data(
            format!("{display} 点歌成功"),
            serde_json::json!({"playback_mode": profile.mode.as_str()}),
        )
    }

    pub async fn list_pending(&self) -> ApiResult {
        let histories = match self.histories.list_pending().await {
            Ok(h) => h,
            Err(e) => return ApiResult::fail(db_error_message(&e, "获取队列失败")),
        };
        let total = histories.len() as i64;
        let data = self.build_list(&histories).await;
        let mut result = ApiResult::ok_with_data(data).with_pagination(total, 1, PAGE_SIZE);
        result.total_page = if total > 0 { 1 } else { 0 };
        result
    }

    pub async fn list_history(&self, page: i64) -> ApiResult {
        let (histories, total) = match self.histories.list_history_page(page).await {
            Ok(v) => v,
            Err(e) => return ApiResult::fail(db_error_message(&e, "获取队列失败")),
        };
        let data = self.build_list(&histories).await;
        ApiResult::ok_with_data(data).with_pagination(total, page.max(1), PAGE_SIZE)
    }

    pub async fn list_usually(&self, page: i64) -> ApiResult {
        let (histories, total) = match self.histories.list_usually_page(page).await {
            Ok(v) => v,
            Err(e) => return ApiResult::fail(db_error_message(&e, "获取队列失败")),
        };
        let data = self.build_list(&histories).await;
        ApiResult::ok_with_data(data).with_pagination(total, page.max(1), PAGE_SIZE)
    }

    pub async fn set_top(&self, song_id: i64) -> ApiResult {
        let history = match self.histories.get_optional(song_id).await {
            Ok(Some(h)) => h,
            Ok(None) => return ApiResult::fail("歌曲不在队列中"),
            Err(e) => return ApiResult::fail(db_error_message(&e, "置顶失败")),
        };
        if let Err(e) = self.histories.set_top(song_id).await {
            return ApiResult::fail(db_error_message(&e, "置顶失败"));
        }
        self.events.publish_queue_changed();
        ApiResult::ok_msg(format!("{} 置顶成功", history.name))
    }

    pub async fn remove_if_exists(&self, song_id: i64) {
        if let Ok(Some(_)) = self.histories.get_optional(song_id).await {
            let _ = self.histories.delete(song_id).await;
        }
    }

    pub async fn remove(&self, song_id: i64) -> ApiResult {
        let history = match self.histories.get_optional(song_id).await {
            Ok(Some(h)) => h,
            Ok(None) => return ApiResult::fail("歌曲不在队列中"),
            Err(e) => return ApiResult::fail(db_error_message(&e, "移除队列项失败")),
        };
        if let Err(e) = self.histories.delete(song_id).await {
            return ApiResult::fail(db_error_message(&e, "移除队列项失败"));
        }
        self.events.publish_queue_changed();
        ApiResult::ok_msg(format!("{} 播放记录删除成功", history.name))
    }
}
